File: mobile/android/app/src/main/java/com/volttracker/obdpoc/materialize/ChargeSessionMaterializer.java
package com.volttracker.obdpoc.materialize;

import java.util.ArrayList;
import java.util.List;

/**
 * Conservative charge-session materializer.
 *
 * <p><b>Primary signal — HV pack current.</b> When {@link TelemetrySample#packCurrentA} is
 * available (signed; discharge positive per the Volt convention) we treat a sample as plugged when
 * current is flowing <em>into</em> the pack — {@code packCurrentA <= -}{@link
 * Tunables#CHARGING_PACK_CURRENT_A_THRESHOLD}. Combined with near-zero speed, that's a
 * high-confidence "the car is on a charger" signal that {@link Confidence#OBSERVED} earns.
 *
 * <p><b>Fallback — aux 12V.</b> When pack current isn't in the sample (older sessions from
 * pre-schema-v8 installs, or vehicles where 222414 doesn't respond), we fall back to the legacy
 * heuristic: {@link TelemetrySample#adapterVoltage} above {@link
 * Tunables#PLUGGED_VOLTAGE_THRESHOLD} with near-zero speed. This is documented {@link
 * Confidence#WEAK} because the alternator under load can raise 12V transiently without an EVSE
 * involved.
 *
 * <p><b>Gap behaviour.</b> A single gap inside an otherwise-plugged window <em>merges</em> into one
 * session with {@link ChargeSession#interruptionCount} incremented; the intent is that a brief
 * cable jiggle or polling stall during a real charge does not split the row. Only gaps that exceed
 * {@link Tunables#SPLIT_GAP_MS} (i.e. plugged signal absent for that long) actually split the
 * window into two sessions.
 */
public final class ChargeSessionMaterializer {

    private ChargeSessionMaterializer() {}

    static final class Tunables {
        /** Brief gap inside a session — counted as an interruption but not a split. */
        static final long MAX_GAP_MS = 5L * 60_000L;

        /** Gap above this splits the window into two separate charge sessions. */
        static final long SPLIT_GAP_MS = 30L * 60_000L;

        /** Below this duration the window is rejected (likely noise). */
        static final long MIN_DURATION_MS = 60_000L;

        /**
         * Pack current more negative than this (i.e. ≥ this many amps flowing INTO the pack) is
         * treated as charging. The 1.0 A floor filters out parked-with-12V-conversion noise (the
         * DC-DC trickle from the pack into the 12V battery can show up as a tiny negative current
         * at idle); real EVSE charging is tens to hundreds of amps.
         */
        static final double CHARGING_PACK_CURRENT_A_THRESHOLD = 1.0;

        /** Adapter voltage above this with low speed indicates the car is on a charger. */
        static final double PLUGGED_VOLTAGE_THRESHOLD = 13.5;

        /** Speed above this means the car is moving — not plugged in. */
        static final double STATIONARY_SPEED_KPH = 1.0;
    }

    public static List<ChargeSession> materialize(MaterializerInput in, MaterializerData data) {
        List<ChargeSession> result = new ArrayList<>();
        if (in == null || data == null) {
            return result;
        }
        List<TelemetrySample> telemetry = data.readTelemetrySamples(in.sessionId);

        if (telemetry == null || telemetry.isEmpty()) {
            // Without telemetry we have no continuous window to anchor a charge session on.
            return result;
        }

        // Walk the samples once to find out whether the high-confidence pack-current signal
        // ever fires in this session; if it does we materialize OBSERVED, otherwise the
        // fallback voltage heuristic produces WEAK rows.
        boolean usedPackCurrent = false;
        List<TelemetrySample> currentRun = new ArrayList<>();
        int interruptions = 0;

        for (TelemetrySample sample : telemetry) {
            PluggedReason plugged = isPluggedSample(sample);
            if (plugged != PluggedReason.NOT_PLUGGED) {
                if (plugged == PluggedReason.PACK_CURRENT) {
                    usedPackCurrent = true;
                }
                if (!currentRun.isEmpty()) {
                    long gap =
                            sample.capturedAtMs
                                    - currentRun.get(currentRun.size() - 1).capturedAtMs;
                    if (gap > Tunables.SPLIT_GAP_MS) {
                        ChargeSession finalized =
                                finalize(currentRun, interruptions, usedPackCurrent);
                        if (finalized != null) {
                            result.add(finalized);
                        }
                        currentRun = new ArrayList<>();
                        interruptions = 0;
                        usedPackCurrent = plugged == PluggedReason.PACK_CURRENT;
                    } else if (gap > Tunables.MAX_GAP_MS) {
                        interruptions += 1;
                    }
                }
                currentRun.add(sample);
            }
            // Non-plugged samples between plugged runs are ignored — only the gap timestamp
            // matters and that is computed off the last plugged sample.
        }
        if (!currentRun.isEmpty()) {
            ChargeSession finalized = finalize(currentRun, interruptions, usedPackCurrent);
            if (finalized != null) {
                result.add(finalized);
            }
        }
        return result;
    }

    /** Why a sample qualified as plugged, or {@link #NOT_PLUGGED} if it didn't. */
    private enum PluggedReason {
        NOT_PLUGGED,
        PACK_CURRENT,
        AUX_VOLTAGE
    }

    /**
     * Combines the primary pack-current signal with the legacy aux-voltage fallback.
     *
     * <p>Discharge is positive in this codebase, so real charging current is <em>negative</em>. A
     * strong negative current (at or beyond {@link Tunables#CHARGING_PACK_CURRENT_A_THRESHOLD} in
     * magnitude) is a high-confidence "the car is on a charger" signal — many amps cannot flow INTO
     * the pack while the car is moving — so it overrides a missing or near-zero speed reading. This
     * matters in practice because the Volt's 0xFF speed sentinel during a charge makes the {@code
     * speed_kph} column NULL for most of the session; without the short-circuit we'd reject every
     * charging sample on the sentinel-mode path.
     *
     * <p>For weaker signals (no pack current at all, or aux voltage only), we still require a valid
     * stationary speed reading because the aux-voltage heuristic is too noisy to trust on its own.
     */
    private static PluggedReason isPluggedSample(TelemetrySample sample) {
        if (sample == null) {
            return PluggedReason.NOT_PLUGGED;
        }
        boolean strongCharging =
                sample.packCurrentA != null
                        && sample.packCurrentA <= -Tunables.CHARGING_PACK_CURRENT_A_THRESHOLD;
        boolean clearlyMoving =
                sample.speedKph != null && sample.speedKph > Tunables.STATIONARY_SPEED_KPH;
        if (clearlyMoving) {
            return PluggedReason.NOT_PLUGGED;
        }
        if (strongCharging) {
            return PluggedReason.PACK_CURRENT;
        }
        if (sample.speedKph == null) {
            // Without a strong pack-current signal AND without a speed reading, we cannot
            // distinguish stationary-plugged from moving-with-regen — be conservative.
            return PluggedReason.NOT_PLUGGED;
        }
        if (sample.packCurrentA != null) {
            // Definitive pack-current reading that does NOT indicate charging — overrides the
            // noisier aux-voltage heuristic.
            return PluggedReason.NOT_PLUGGED;
        }
        if (sample.adapterVoltage != null
                && sample.adapterVoltage > Tunables.PLUGGED_VOLTAGE_THRESHOLD) {
            return PluggedReason.AUX_VOLTAGE;
        }
        return PluggedReason.NOT_PLUGGED;
    }

    private static ChargeSession finalize(
            List<TelemetrySample> run, int interruptions, boolean usedPackCurrent) {
        if (run.isEmpty()) {
            return null;
        }
        long startedAtMs = run.get(0).capturedAtMs;
        long endedAtMs = run.get(run.size() - 1).capturedAtMs;
        long durationMs = Math.max(0L, endedAtMs - startedAtMs);
        if (durationMs < Tunables.MIN_DURATION_MS) {
            return null;
        }
        Double voltageStart = run.get(0).adapterVoltage;
        Double voltageEnd = run.get(run.size() - 1).adapterVoltage;
        Confidence confidence = usedPackCurrent ? Confidence.OBSERVED : Confidence.WEAK;
        Double startSoc = firstFiniteSoc(run);
        Double endSoc = lastFiniteSoc(run);
        Double peakPowerKw = peakChargePowerKw(run);
        Double energyKwh = integrateChargeEnergyKwh(run);
        return new ChargeSession(
                startedAtMs,
                endedAtMs,
                durationMs,
                voltageStart,
                voltageEnd,
                interruptions,
                "unknown",
                confidence,
                startSoc,
                endSoc,
                peakPowerKw,
                energyKwh);
    }

    private static Double firstFiniteSoc(List<TelemetrySample> run) {
        for (TelemetrySample s : run) {
            if (s.socPct != null && !s.socPct.isNaN() && !s.socPct.isInfinite()) {
                return s.socPct;
            }
        }
        return null;
    }

    private static Double lastFiniteSoc(List<TelemetrySample> run) {
        for (int i = run.size() - 1; i >= 0; i--) {
            TelemetrySample s = run.get(i);
            if (s.socPct != null && !s.socPct.isNaN() && !s.socPct.isInfinite()) {
                return s.socPct;
            }
        }
        return null;
    }

    /**
     * Peak power INTO the pack during the window, in kW. Pack convention is discharge-positive, so
     * charging power = {@code -(packVoltage * packCurrentA / 1000)}. Returns {@code null} when no
     * sample in the run carried both values.
     */
    private static Double peakChargePowerKw(List<TelemetrySample> run) {
        Double peak = null;
        for (TelemetrySample s : run) {
            if (s.packVoltage == null || s.packCurrentA == null) {
                continue;
            }
            double kw = -(s.packVoltage * s.packCurrentA) / 1000.0;
            if (Double.isNaN(kw) || Double.isInfinite(kw)) {
                continue;
            }
            if (kw <= 0.0) {
                continue;
            }
            if (peak == null || kw > peak) {
                peak = kw;
            }
        }
        return peak;
    }

    /**
     * Trapezoidal integration of charging power across the run. Same sign convention as {@link
     * #peakChargePowerKw}. Returns {@code null} when fewer than two samples in the window carry
     * both pack values (a single point can't form an area).
     */
    private static Double integrateChargeEnergyKwh(List<TelemetrySample> run) {
        Double prevKw = null;
        Long prevMs = null;
        double energyKwh = 0.0;
        int integrated = 0;
        for (TelemetrySample s : run) {
            if (s.packVoltage == null || s.packCurrentA == null) {
                continue;
            }
            double kw = -(s.packVoltage * s.packCurrentA) / 1000.0;
            if (Double.isNaN(kw) || Double.isInfinite(kw)) {
                continue;
            }
            // Clip negative (discharge) samples to 0 so a brief discharge dip inside an otherwise-
            // plugged window doesn't subtract from the integrated charge total.
            if (kw < 0.0) {
                kw = 0.0;
            }
            if (prevKw != null && prevMs != null) {
                double hours = (s.capturedAtMs - prevMs) / 3_600_000.0;
                if (hours > 0.0) {
                    energyKwh += ((prevKw + kw) / 2.0) * hours;
                    integrated += 1;
                }
            }
            prevKw = kw;
            prevMs = s.capturedAtMs;
        }
        return integrated == 0 ? null : energyKwh;
    }
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/materialize/ChargeSessionMaterializer.kt
package com.volttracker.obdpoc.materialize

/**
 * Conservative charge-session materializer.
 *
 * **Primary signal — HV pack current.** When [TelemetrySample.packCurrentA] is available (signed;
 * discharge positive per the Volt convention) we treat a sample as plugged when current is flowing
 * *into* the pack — `packCurrentA <= -`[Tunables.CHARGING_PACK_CURRENT_A_THRESHOLD]. Combined with
 * near-zero speed, that's a high-confidence "the car is on a charger" signal that
 * [Confidence.OBSERVED] earns.
 *
 * **Fallback — aux 12V.** When pack current isn't in the sample we fall back to the legacy
 * heuristic: [TelemetrySample.adapterVoltage] above [Tunables.PLUGGED_VOLTAGE_THRESHOLD] with
 * near-zero speed. This is documented [Confidence.WEAK] because the alternator under load can raise
 * 12V transiently without an EVSE involved.
 *
 * **Gap behaviour.** A single gap inside an otherwise-plugged window *merges* into one session with
 * [ChargeSession.interruptionCount] incremented; the intent is that a brief cable jiggle or polling
 * stall during a real charge does not split the row. Only gaps that exceed [Tunables.SPLIT_GAP_MS]
 * actually split the window into two sessions.
 *
 * **Transient-break debounce.** A momentary movement/discharge sample (a one-off GPS speed glitch, a
 * DC-DC converter blip, a brief power spike) no longer instantly splits an active charge — that was
 * the main cause of one physical charge being logged as several sessions. A break only splits the
 * window once it is *sustained* for [Tunables.BREAK_DEBOUNCE_MS]; a shorter break that resumes
 * charging is folded back in as an [ChargeSession.interruptionCount]. A genuinely sustained drive
 * still splits, so a drive is never stitched into a charge. A finalized window also needs at least
 * [Tunables.MIN_SAMPLES] plugged samples, which rejects sparse two-sample noise.
 */
object ChargeSessionMaterializer {
    /**
     * Charge-session thresholds are pinned to the policy in ADR 0004.
     *
     * The key calibration is the preference for Volt pack current over aux-voltage inference:
     * negative pack current is high-confidence charging evidence, while `ATRV` above 13.5 V is
     * only a fallback because post-drive 12V behavior can look charger-like. The split/merge
     * windows protect short cable wiggles or polling stalls without joining genuinely separate
     * charge sessions. See `docs/adr/0004-charge-detection-heuristics.md` and the materializer
     * tests before changing these values.
     */
    private object Tunables {
        /** Brief gap inside a session — counted as an interruption but not a split. */
        const val MAX_GAP_MS = 5L * 60_000L

        /** Gap above this splits the window into two separate charge sessions. */
        const val SPLIT_GAP_MS = 30L * 60_000L

        /** Below this duration the window is rejected (likely noise). */
        const val MIN_DURATION_MS = 60_000L

        /**
         * A break (movement/discharge) must persist at least this long before it splits an active
         * charge. Shorter breaks that resume charging are treated as interruptions, not splits —
         * this is what stops a single glitchy sample from over-counting one charge as several.
         */
        const val BREAK_DEBOUNCE_MS = 60_000L

        /**
         * A finalized window needs at least this many plugged samples. Real charges stream hundreds
         * of samples; this floor rejects the sparse two-sample windows that a transient break can
         * otherwise leave behind.
         */
        const val MIN_SAMPLES = 3

        /**
         * Pack current more negative than this (i.e. ≥ this many amps flowing INTO the pack) is
         * treated as charging. The 1.0 A floor filters out parked-with-12V-conversion noise; real
         * EVSE charging is tens to hundreds of amps.
         */
        const val CHARGING_PACK_CURRENT_A_THRESHOLD = 1.0

        /** Adapter voltage above this with low speed indicates the car is on a charger. */
        const val PLUGGED_VOLTAGE_THRESHOLD = 13.5

        /** Speed above this means the car is moving — not plugged in. */
        const val STATIONARY_SPEED_KPH = 1.0
    }

    @JvmStatic
    fun materialize(
        input: MaterializerInput?,
        data: MaterializerData?,
    ): List<ChargeSession> {
        val result = ArrayList<ChargeSession>()
        if (input == null || data == null) {
            return result
        }
        val telemetry = data.readTelemetrySamples(input.sessionId)
        if (telemetry.isEmpty()) {
            // Without telemetry we have no continuous window to anchor a charge session on.
            return result
        }

        // Walk the samples once, tracking per run whether the high-confidence pack-current signal
        // ever fires inside it; if it does that run materializes OBSERVED, otherwise the fallback
        // voltage heuristic produces a WEAK row.
        var usedPackCurrent = false
        var currentRun = ArrayList<TelemetrySample>()
        var interruptions = 0
        // Timestamp of the first breaking sample of an as-yet-unresolved break, or null when the
        // run is not currently interrupted. Drives the transient-break debounce below.
        var pendingBreakStartMs: Long? = null

        for (sample in telemetry) {
            val plugged = isPluggedSample(sample)
            if (plugged != PluggedReason.NOT_PLUGGED) {
                if (currentRun.isNotEmpty()) {
                    val breakStartMs = pendingBreakStartMs
                    val hadTransientBreak = breakStartMs != null
                    // With sparse telemetry there may be only one breaking row. Re-check the break
                    // duration when charging resumes; otherwise a multi-minute drive is treated as
                    // a one-sample blip and two physical charges are stitched together.
                    val sustainedBreak =
                        breakStartMs != null && sample.capturedAtMs - breakStartMs > Tunables.BREAK_DEBOUNCE_MS
                    pendingBreakStartMs = null
                    val gap = sample.capturedAtMs - currentRun[currentRun.size - 1].capturedAtMs
                    if (gap > Tunables.SPLIT_GAP_MS || sustainedBreak) {
                        // This boundary is a SPLIT, not an interruption — so don't count the pending
                        // transient break against the run being closed. Finalize with the flag state
                        // as of ITS OWN samples; the incoming sample's evidence belongs to the run
                        // that starts with it.
                        val finalized = finalizeRun(currentRun, interruptions, usedPackCurrent)
                        if (finalized != null) {
                            result.add(finalized)
                        }
                        currentRun = ArrayList()
                        interruptions = 0
                        usedPackCurrent = false
                    } else if (hadTransientBreak || gap > Tunables.MAX_GAP_MS) {
                        // The run continues: a transient break (movement/discharge that resumed
                        // before BREAK_DEBOUNCE_MS) or a moderate gap counts as exactly one
                        // interruption, never both for the same resume.
                        interruptions += 1
                    }
                }
                if (plugged == PluggedReason.PACK_CURRENT) {
                    usedPackCurrent = true
                }
                currentRun.add(sample)
            } else if (currentRun.isNotEmpty() && breaksChargeRun(sample)) {
                val breakStart = pendingBreakStartMs ?: sample.capturedAtMs
                pendingBreakStartMs = breakStart
                if (sample.capturedAtMs - breakStart >= Tunables.BREAK_DEBOUNCE_MS) {
                    // Sustained movement/discharge: a real drive (or an unplug-and-leave), so split.
                    val finalized = finalizeRun(currentRun, interruptions, usedPackCurrent)
                    if (finalized != null) {
                        result.add(finalized)
                    }
                    currentRun = ArrayList()
                    interruptions = 0
                    usedPackCurrent = false
                    pendingBreakStartMs = null
                }
            }
            // Non-plugged samples between plugged runs are ignored — only the gap timestamp matters
            // and that is computed off the last plugged sample. Sustained movement/discharge is the
            // exception: it breaks the candidate so a drive cannot be stitched into a charge.
        }
        if (currentRun.isNotEmpty()) {
            val finalized = finalizeRun(currentRun, interruptions, usedPackCurrent)
            if (finalized != null) {
                result.add(finalized)
            }
        }
        return result
    }

    /** Why a sample qualified as plugged, or [NOT_PLUGGED] if it didn't. */
    private enum class PluggedReason {
        NOT_PLUGGED,
        PACK_CURRENT,
        AUX_VOLTAGE,
    }

    /**
     * Combines the primary pack-current signal with the legacy aux-voltage fallback. Discharge is
     * positive in this codebase, so real charging current is *negative*. A strong negative current
     * overrides a missing/near-zero speed reading (the Volt's 0xFF speed sentinel during a charge
     * makes `speed_kph` NULL for most of the session). For weaker signals we still require a valid
     * stationary speed reading because the aux-voltage heuristic is too noisy to trust on its own.
     */
    private fun isPluggedSample(sample: TelemetrySample?): PluggedReason {
        if (sample == null) {
            return PluggedReason.NOT_PLUGGED
        }
        val packCurrent = sample.packCurrentA
        val speed = sample.speedKph
        val strongCharging =
            packCurrent != null &&
                packCurrent.isFinite() &&
                packCurrent <= -Tunables.CHARGING_PACK_CURRENT_A_THRESHOLD
        val clearlyMoving = speed != null && speed.isFinite() && speed > Tunables.STATIONARY_SPEED_KPH
        if (clearlyMoving) {
            return PluggedReason.NOT_PLUGGED
        }
        if (strongCharging) {
            return PluggedReason.PACK_CURRENT
        }
        if (speed == null || !speed.isFinite()) {
            // Without a strong pack-current signal AND without a speed reading, we cannot
            // distinguish stationary-plugged from moving-with-regen — be conservative.
            return PluggedReason.NOT_PLUGGED
        }
        if (packCurrent != null) {
            // Definitive pack-current reading that does NOT indicate charging — overrides the
            // noisier aux-voltage heuristic.
            return PluggedReason.NOT_PLUGGED
        }
        val voltage = sample.adapterVoltage
        if (voltage != null && voltage.isFinite() && voltage > Tunables.PLUGGED_VOLTAGE_THRESHOLD) {
            return PluggedReason.AUX_VOLTAGE
        }
        return PluggedReason.NOT_PLUGGED
    }

    private fun breaksChargeRun(sample: TelemetrySample?): Boolean {
        if (sample == null) {
            return false
        }
        val speed = sample.speedKph
        if (speed != null && speed.isFinite() && speed > Tunables.STATIONARY_SPEED_KPH) {
            return true
        }
        val packCurrent = sample.packCurrentA
        if (packCurrent != null && packCurrent.isFinite() && packCurrent > Tunables.CHARGING_PACK_CURRENT_A_THRESHOLD) {
            return true
        }
        val power = sample.powerKw
        return power != null && power.isFinite() && power > 1.0
    }

    private fun finalizeRun(
        run: List<TelemetrySample>,
        interruptions: Int,
        usedPackCurrent: Boolean,
    ): ChargeSession? {
        if (run.isEmpty()) {
            return null
        }
        if (run.size < Tunables.MIN_SAMPLES) {
            // Too few plugged samples to trust — a sparse two-sample window is noise, not a charge.
            return null
        }
        val startedAtMs = run[0].capturedAtMs
        val endedAtMs = run[run.size - 1].capturedAtMs
        val durationMs = maxOf(0L, endedAtMs - startedAtMs)
        if (durationMs < Tunables.MIN_DURATION_MS) {
            return null
        }
        val voltageStart = run[0].adapterVoltage?.takeIf { it.isFinite() }
        val voltageEnd = run[run.size - 1].adapterVoltage?.takeIf { it.isFinite() }
        val confidence = if (usedPackCurrent) Confidence.OBSERVED else Confidence.WEAK
        return ChargeSession(
            startedAtMs,
            endedAtMs,
            durationMs,
            voltageStart,
            voltageEnd,
            interruptions,
            "unknown",
            confidence,
            firstFiniteSoc(run),
            lastFiniteSoc(run),
            peakChargePowerKw(run),
            integrateChargeEnergyKwh(run),
        )
    }

    private fun firstFiniteSoc(run: List<TelemetrySample>): Double? {
        for (s in run) {
            val soc = s.socPct
            if (soc != null && !soc.isNaN() && !soc.isInfinite()) {
                return soc
            }
        }
        return null
    }

    private fun lastFiniteSoc(run: List<TelemetrySample>): Double? {
        for (i in run.indices.reversed()) {
            val soc = run[i].socPct
            if (soc != null && !soc.isNaN() && !soc.isInfinite()) {
                return soc
            }
        }
        return null
    }

    /**
     * Peak power INTO the pack during the window, in kW. Pack convention is discharge-positive, so
     * charging power = `-(packVoltage * packCurrentA / 1000)`. Returns `null` when no sample in the
     * run carried both values.
     */
    private fun peakChargePowerKw(run: List<TelemetrySample>): Double? {
        var peak: Double? = null
        for (s in run) {
            val v = s.packVoltage
            val c = s.packCurrentA
            if (v == null || c == null) {
                continue
            }
            val kw = PackEnergyMath.chargePowerKw(v, c) ?: continue
            if (kw <= 0.0) {
                continue
            }
            if (peak == null || kw > peak) {
                peak = kw
            }
        }
        return peak
    }

    /**
     * Trapezoidal integration of charging power across the run. Same sign convention as
     * [peakChargePowerKw]. Returns `null` when fewer than two samples in the window carry both pack
     * values (a single point can't form an area).
     */
    private fun integrateChargeEnergyKwh(run: List<TelemetrySample>): Double? {
        var prevKw: Double? = null
        var prevMs: Long? = null
        var energyKwh = 0.0
        var integrated = 0
        var lastV: Double? = null
        for (s in run) {
            val c = s.packCurrentA ?: continue
            // Pack voltage is polled on a slower lane than current, so carry the last-known voltage
            // across a current-only sample rather than dropping it from the integral (B2).
            s.packVoltage?.let { lastV = it }
            val v = lastV ?: continue
            var kw = PackEnergyMath.chargePowerKw(v, c) ?: continue
            // Clip negative (discharge) samples to 0 so a brief discharge dip inside an otherwise-
            // plugged window doesn't subtract from the integrated charge total.
            if (kw < 0.0) {
                kw = 0.0
            }
            val pKw = prevKw
            val pMs = prevMs
            if (pKw != null && pMs != null) {
                val segmentKwh = PackEnergyMath.trapezoidKwh(pKw, kw, pMs, s.capturedAtMs)
                if (segmentKwh != null) {
                    energyKwh += segmentKwh
                    integrated += 1
                }
            }
            prevKw = kw
            prevMs = s.capturedAtMs
        }
        return if (integrated == 0) null else energyKwh
    }
}
